### ISSUE TITLE
perf: bulk-insert fast-path

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#6373]
From discussion yesterday regarding optimizing bulk inserts, we should create a fast-path when a single `BatchRequest` contains multiple `Put` or `ConditionalPut` calls. Currently, each `{Put,ConditionalPut}` request results in a call to `MVCC{Put,ConditionalPut}` which in turn call down to `mvccPutInternal`. This function is general purpose and checks whether a version of the key being put already exists. But in bulk-insert scenarios we could perform a pre-check involving a single iterator seek to check whether a whole range of keys exist. 

At the mvcc layer, we could create `MVCCBlind{Put,ConditionalPut}` operations. A simple experiment to disable the creation and usage of the iterator in benchmark scenarios shows these would be significantly faster than the more general `MVCC{Put,ConditionalPut}`:

```
name                            old time/op    new time/op     delta
MVCCPut10_RocksDB-8               3.87µs ± 2%     1.40µs ± 2%   -63.80%  (p=0.008 n=5+5)
MVCCPut100_RocksDB-8              3.93µs ± 3%     1.57µs ± 7%   -60.16%  (p=0.008 n=5+5)
MVCCPut1000_RocksDB-8             5.98µs ± 4%     2.96µs ± 4%   -50.53%  (p=0.008 n=5+5)
MVCCPut10000_RocksDB-8            22.1µs ± 1%     17.3µs ± 1%   -21.82%  (p=0.008 n=5+5)
MVCCBatch1Put10_RocksDB-8         5.60µs ± 5%     2.49µs ± 3%   -55.48%  (p=0.008 n=5+5)
MVCCBatch100Put10_RocksDB-8       3.62µs ± 4%     0.75µs ± 8%   -79.25%  (p=0.008 n=5+5)
MVCCBatch10000Put10_RocksDB-8     3.83µs ± 8%     0.72µs ± 2%   -81.31%  (p=0.008 n=5+5)
MVCCBatch100000Put10_RocksDB-8    3.85µs ± 5%     0.72µs ± 3%   -81.22%  (p=0.008 n=5+5)

name                            old speed      new speed       delta
MVCCPut10_RocksDB-8             2.58MB/s ± 1%   7.14MB/s ± 2%  +176.24%  (p=0.008 n=5+5)
MVCCPut100_RocksDB-8            25.5MB/s ± 3%   64.0MB/s ± 7%  +151.28%  (p=0.008 n=5+5)
MVCCPut1000_RocksDB-8            167MB/s ± 4%    338MB/s ± 3%  +102.12%  (p=0.008 n=5+5)
MVCCPut10000_RocksDB-8           452MB/s ± 1%    579MB/s ± 1%   +27.91%  (p=0.008 n=5+5)
MVCCBatch1Put10_RocksDB-8       1.79MB/s ± 6%   4.01MB/s ± 2%  +124.27%  (p=0.008 n=5+5)
MVCCBatch100Put10_RocksDB-8     2.76MB/s ± 4%  13.33MB/s ± 7%  +382.42%  (p=0.008 n=5+5)
MVCCBatch10000Put10_RocksDB-8   2.61MB/s ± 7%  13.96MB/s ± 2%  +434.20%  (p=0.008 n=5+5)
MVCCBatch100000Put10_RocksDB-8  2.60MB/s ± 5%  13.82MB/s ± 3%  +432.10%  (p=0.008 n=5+5)
```

In order to use these new operations, `Replica.executeBatch` would be enhanced to gather up a contiguous run of `Put` and `ConditionalPut` operations and then group them by key ranges which are determined to be empty.

**Comment by tschottdorf** [comment]
Somewhat unrelated, but are there bullet points on the bulk insert discussion? Is this a significant low-hanging fruit? From #5981 it seems like there are plenty more, so I'm curious what other changes to the pipeline were discussed.

**Comment by petermattis** [comment]
This is the most significant low-hanging fruit. There is some additional optimization work that can be done at the SQL layer to handle bulk inserts better, but that will be a much smaller win. The other big win would be to implement something like [column groups](https://github.com/cockroachdb/cockroach/pull/3435).

**Comment by tschottdorf** [comment]
Are you sure? Maybe from the POV of the MVCC-level benchmark, but I thought the command queue was accounting for a lot of it as well. Was there any discussion of features upstream of MVCC?

**Comment by petermattis** [comment]
Am I sure this is the most significant low-hanging fruit? No. This just doesn't seem that difficult to fix. Investigating the command queue latency should certainly be done as well. I might have overstated the scope of the discussion: there was an observation that bulk-insert could benefit from a fast-path `MVCC{Put,ConditionalPut}`.

**Comment by petermattis** [comment]
Implemented by #6414.
